### PR TITLE
RosBag odd padding parsing error fix

### PR DESCRIPTION
**Pull request by brendangeck**
Description of Changes
======================
Currently, we naively iterate through the bagfile records, parsing them one-by-one, assuming the next record always ends flush to the previous record. In some very rare cases, a bagfile will contain junk/padding between the end of the `CHUNK` and `INDEX` sections of the bag, breaking the previously stated assumption.

This PR modifies how we parse bags, using a more informed models, where we seek to specific locations in the file that we are certain contain specific record types. This is the same core algorithm as what is used in the official `rospy` project: https://github.com/ros/ros_comm/blob/noetic-devel/tools/rosbag/src/rosbag/bag.py#L2638-L2668

Meanwhile, I also added `gdb` to the Dockerfile to easier make use of the `--compilation_mode dbg` when performing a `bazel build`.

### CODE REVIEW

**Comment by liamembark** [comment]
Awesome stuff! Would be interested to know how this affects performance at all. Also, I'm not what you would call a good C++ developer, so would like to get some insight on the code from @jaredjxyz or @zig-for. Will look over the general concept though!

File: lib/embag.cc [hunk 1]
```diff
@@ -147,138 +147,152 @@ RosBagTypes::header_t Bag::readHeader(const RosBagTypes::record_t &record) {
...
+    chunk_header.getField("compression", chunk.compression);
+    chunk_header.getField("size", chunk.uncompressed_size);
+
+    if (!(chunk.compression == "lz4" || chunk.compression == "bz2" || chunk.compression == "none")) {
+      throw std::runtime_error("Unsupported compression type: " + chunk.compression);
+    }
+
+    chunk.info = info;
+
+    chunks_.emplace_back(chunk);
```
Comment (jaredjxyz) [inline lib/embag.cc:273]: Nit: emplace_back probably works here, but push_back might be more semantically appropriate
https://abseil.io/tips/112

File: lib/embag.cc [hunk 2]
```diff
@@ -147,138 +147,152 @@ RosBagTypes::header_t Bag::readHeader(const RosBagTypes::record_t &record) {
...
+    index_data_header.getField("count", msg_count);
+
+    RosBagTypes::index_block_t index_block{};
+
+    // NOTE: It seems like it would be simpler to just do &chunk here right? WRONG.
+    //       C++ resuses the same memory location for the chunk variable for each loop, so
+    //       if you use &chunk, all `into_chunk` values will be exactly the same
+    index_block.into_chunk = &chunks_[i];
+
+    connections_[connection_id].blocks.emplace_back(index_block);
```
Comment (jaredjxyz) [inline lib/embag.cc:295]: Same nit here

**Review by jaredjxyz — COMMENTED** [review]
In general I see no issue here, code is pretty clean and good comments, not too hard to follow along with what rospy does, thanks.

I would really like to see a well thought out testing plan - Can you make sure to test this pretty extensively to ensure it works in all places we're using embag? It could be a pretty big issue if there's a bug in here.

File: lib/embag.cc (hunk 1 above)
Comment (brendangeck) [reply lib/embag.cc:273]: Jason has used `emplace_back` throughout the vast majority of embag, so I wanted to keep it consistent, and I felt that `emplace` didn't really cause much confusion here, as the type of `chunk` is clear.

Comment (jaredjxyz) [reply lib/embag.cc:273]: That sounds like consistency for the sake of consistency to me, not consistency for the sake of clarity. If you're going to continue a negative trend, I think you should have a pretty good reason to do it.

When reading this code, I had to take a little extra time to figure out why you were emplacing the object instead of pushing it back. Emplace has a specific meaning, and misusing it can (and, in my opinion, has) make it extra effort for the reader.

Don't be afraid to make improvements and improve the codebase in order to improve the direction of a codebase when there is a negative trend. Jason didn't know the difference between push_back and emplace_back when he wrote this - I know, because I asked him.

Comment (brendangeck) [reply lib/embag.cc:273]: Okay I buy that

**Comment by liamembark** [comment]
Fixes: DATA-418
